File: tools/tal_support/as274_fc/F/t9.f
	program wampler
c
c	Test of includ, tolset, sing and regcf, using the 10x9 test
c	matrices from:
c	Wampler, R.H. (1980) Test procedures and test problems for
c	     least squares algorithms. J. of Econometrics, 12, 3-22.
c
	double precision weight, xrow(9), y, d(9), rbar(36)
	double precision thetab(9), sserr,
     +		tol(9), work(9), beta(9), zero, one, pt1, nine, eps
	integer np, nrbar, ier, k, i, j
	logical lindep(9)
	data weight/1.0/, zero/0.0/, one/1.0/, pt1/0.1/, nine/9.0/,
     +          np/9/, nrbar/36/
c
c       Single precision, starting eps = 0.1
c       Double precision, starting eps = 1.d-05
c
	eps = 1.0d-5
c
	do 50 k = 1, 5
	  call clear(np, nrbar, d, rbar, thetab, sserr, ier)
	  do 10 j = 1, 9
   10	  xrow(j) = one
	  y = nine + eps
	  call includ(np, nrbar, weight, xrow, y, d, rbar, thetab,
     +          sserr, ier)
	  y = eps
	  do 30 i = 2, 10
	    do 20 j = 2, 9
   20       xrow(j) = zero
	    xrow(1) = one
	    if (i .lt. 10) xrow(i) = eps
	    call includ(np, nrbar, weight, xrow, y, d, rbar, thetab,
     +		sserr, ier)
	    if (i .eq. 9) y = nine - eps
   30	  continue
c
c	Set tol & then test for singularities
c
	  call tolset(np, nrbar, d, rbar, tol, work, ier)
	  call sing(np, nrbar, d, rbar, thetab, sserr, tol, lindep,
     +          work, ier)
	  write(*, *)'eps = ', eps
c
c	Calculate regression coefficients if non-singular
c
	  if (ier .eq. 0) then
	    call regcf(np, nrbar, d, rbar, thetab, tol, beta, np, ier)
	    do 40 i = 1, 9
   40	    work(i) = beta(i) - one
	    write(*, 900)(work(i),i=1,9)
  900       format(' Errors in regression coefficients:', 2(/1x,5e11.3))
	  else
	    write(*, *)'ier = ', ier
	  end if
	  write(*, *)
	  eps = eps * pt1
   50	continue
	end

File: tools/tal_support/as274_fc/C/t9.c
#include <stdio.h>
#include "as274c.h"
#include "misc.h"

int main()
{
    double *xrow, y, *d, *rbar, *thetab, sserr, *tol, *work, *beta;
    double pt1=.1, eps;
    int error, k, i, j, *lindep;

    lindep = ivector(0, 8);
    xrow = dvector(0, 8);
    d = dvector(0, 8);
    rbar = dvector(0, 35);
    thetab = dvector(0, 8);
    tol = dvector(0, 8);
    work = dvector(0, 8);
    beta = dvector(0, 8);

    eps = 1e-5;
    for (k=0; k<5; k++) {
        error = clear(9, 36, d, rbar, thetab, &sserr);
        for (j=0; j<9; j++) xrow[j] = 1;
        y = 9 + eps;
        error = includ(9, 36, 1, xrow, y, d, rbar, thetab, &sserr);
        y = eps;
        for (i=1; i<10; i++) {
            for (j=1; j<9; j++) xrow[j] = 0;
            xrow[0] = 1;
            if (i < 9) xrow[i] = eps;
            error = includ(9, 36, 1, xrow, y, d, rbar, thetab, &sserr);
            if (i == 8) y = 9 - eps;
        }

        /* set tol and look for singularities */
        error = tolset(9, 36, d, rbar, tol);
        error = sing(9, 36, d, rbar, thetab, &sserr, tol, lindep);
        printf("eps = %g\n", eps);

        if (error) printf("Error = %d\n", error);
        else {
            error = regcf(9, 36, d, rbar, thetab, tol, beta, 9);
            for (i=0; i<9; i++) work[i] = beta[i] - 1;
            printf("Errors in regression coefficients:\n");
            putdvec("work ", work, 0, 8);
        }

        eps *= pt1;
    }
    return 0;
}
